Add the ability to disable the conf.d inclusion through the node wizard

mcktr opened the pull request:
This PR add the ability to disable the conf.d inclusion through running node wizard. This PR also adds a new function to NodeUtility to update the Icinga 2 configuraton file (`icinga2.conf`). 

```
# /tmp/icinga2/sbin/icinga2 daemon -C -DRunAsUser=$(id -u -n) -DRunAsGroup=$(id -g -n)

information/cli: Icinga application loader (version: v2.6.3-404-ge50a823)
information/cli: Loading configuration file(s).
information/ConfigItem: Committing config item(s).
warning/ApplyRule: Apply rule 'satellite-host' (in /tmp/icinga2/etc/icinga2/conf.d/satellite.conf: 29:1-29:41) for type 'Dependency' does not match anywhere!
information/ConfigItem: Instantiated 3 Zones.
information/ConfigItem: Instantiated 1 FileLogger.
information/ConfigItem: Instantiated 1 Endpoint.
information/ConfigItem: Instantiated 2 NotificationCommands.
information/ConfigItem: Instantiated 12 Notifications.
information/ConfigItem: Instantiated 207 CheckCommands.
information/ConfigItem: Instantiated 1 Host.
information/ConfigItem: Instantiated 1 IcingaApplication.
information/ConfigItem: Instantiated 2 HostGroups.
information/ConfigItem: Instantiated 1 UserGroup.
information/ConfigItem: Instantiated 3 TimePeriods.
information/ConfigItem: Instantiated 1 User.
information/ConfigItem: Instantiated 11 Services.
information/ConfigItem: Instantiated 3 ServiceGroups.
information/ConfigItem: Instantiated 1 ScheduledDowntime.
information/ConfigItem: Instantiated 1 NotificationComponent.
information/ConfigItem: Instantiated 1 CheckerComponent.
information/ScriptGlobal: Dumping variables to file '/tmp/icinga2/var/cache/icinga2/icinga2.vars'
information/cli: Finished validating the configuration file(s).

# /tmp/icinga2/sbin/icinga2 node wizard -DRunAsUser=$(id -u -n) -DRunAsGroup=$(id -g -n)

Welcome to the Icinga 2 Setup Wizard!

We'll guide you through all required configuration details.



Please specify if this is a satellite setup ('n' installs a master setup) [Y/n]: n
Starting the Master setup routine...
Please specify the common name (CN) [ganymed]: 
Do you want do disable the inclusion of the conf.d directory [Y/n]: 
Checking for existing certificates for common name 'ganymed'...
Certificates not yet generated. Running 'api setup' now.
information/cli: Generating new CA.
information/base: Writing private key to '/tmp/icinga2/var/lib/icinga2/ca/ca.key'.
information/base: Writing X509 certificate to '/tmp/icinga2/var/lib/icinga2/ca/ca.crt'.
information/cli: Generating new CSR in '/tmp/icinga2/etc/icinga2/pki/ganymed.csr'.
information/base: Writing private key to '/tmp/icinga2/etc/icinga2/pki/ganymed.key'.
information/base: Writing certificate signing request to '/tmp/icinga2/etc/icinga2/pki/ganymed.csr'.
information/cli: Signing CSR with CA and writing certificate to '/tmp/icinga2/etc/icinga2/pki/ganymed.crt'.
information/pki: Writing certificate to file '/tmp/icinga2/etc/icinga2/pki/ganymed.crt'.
information/cli: Copying CA certificate to '/tmp/icinga2/etc/icinga2/pki/ca.crt'.
Generating master configuration for Icinga 2.
information/cli: Adding new ApiUser 'root' in '/tmp/icinga2/etc/icinga2/conf.d/api-users.conf'.
information/cli: Enabling the 'api' feature.
Enabling feature api. Make sure to restart Icinga 2 for these changes to take effect.
information/cli: Dumping config items to file '/tmp/icinga2/etc/icinga2/zones.conf'.
information/cli: Created backup file '/tmp/icinga2/etc/icinga2/zones.conf.orig'.
Please specify the API bind host/port (optional):
Bind Host []: 
Bind Port []: 
information/cli: Created backup file '/tmp/icinga2/etc/icinga2/features-available/api.conf.orig'.
information/cli: Updating constants.conf.
information/cli: Created backup file '/tmp/icinga2/etc/icinga2/constants.conf.orig'.
information/cli: Updating constants file '/tmp/icinga2/etc/icinga2/constants.conf'.
information/cli: Updating constants file '/tmp/icinga2/etc/icinga2/constants.conf'.
information/cli: Updating constants file '/tmp/icinga2/etc/icinga2/constants.conf'.
information/cli: Updating configuration file '/tmp/icinga2/etc/icinga2/icinga2.conf'.
information/cli: Updating configuration file '/tmp/icinga2/etc/icinga2/icinga2.conf'.
Done.

Now restart your Icinga 2 daemon to finish the installation!

# /tmp/icinga2/sbin/icinga2 daemon -C -DRunAsUser=$(id -u -n) -DRunAsGroup=$(id -g -n)

information/cli: Icinga application loader (version: v2.6.3-404-ge50a823)
information/cli: Loading configuration file(s).
information/ConfigItem: Committing config item(s).
information/ApiListener: My API identity: ganymed
information/ConfigItem: Instantiated 1 ApiUser.
information/ConfigItem: Instantiated 1 ApiListener.
information/ConfigItem: Instantiated 1 Zone.
information/ConfigItem: Instantiated 1 FileLogger.
information/ConfigItem: Instantiated 1 Endpoint.
information/ConfigItem: Instantiated 207 CheckCommands.
information/ConfigItem: Instantiated 1 IcingaApplication.
information/ConfigItem: Instantiated 1 NotificationComponent.
information/ConfigItem: Instantiated 1 CheckerComponent.
information/ScriptGlobal: Dumping variables to file '/tmp/icinga2/var/cache/icinga2/icinga2.vars'
information/cli: Finished validating the configuration file(s).
```

This is still WIP and works so far only while setting up a master instance. Please provide me some feedback if the patch is okay (so far). 

I am trying to add the missing parts within the coming week, so the PR should leave WIP towards the end of coming week.

fixes #4508

mcktr commented:
Oh dear I failed and mixed a bit up within git, so that here are some more commits auto referenced than the branch actually have. Sorry for this :(

The default for disable the inclusion of the conf.d directory is false, so the user must set it explicitly to true (by enter `y` while the node wizard is running).  

```
# /tmp/icinga2/sbin/icinga2 node wizard -DRunAsUser=$(id -u -n) -DRunAsGroup=$(id -g -n)

Welcome to the Icinga 2 Setup Wizard!

We'll guide you through all required configuration details.



Please specify if this is a satellite setup ('n' installs a master setup) [Y/n]: n
Starting the Master setup routine...
Please specify the common name (CN) [ganymed]: 
Checking for existing certificates for common name 'ganymed'...
Certificates not yet generated. Running 'api setup' now.
information/cli: Generating new CA.
information/base: Writing private key to '/tmp/icinga2/var/lib/icinga2/ca/ca.key'.
information/base: Writing X509 certificate to '/tmp/icinga2/var/lib/icinga2/ca/ca.crt'.
information/cli: Generating new CSR in '/tmp/icinga2/etc/icinga2/pki/ganymed.csr'.
information/base: Writing private key to '/tmp/icinga2/etc/icinga2/pki/ganymed.key'.
information/base: Writing certificate signing request to '/tmp/icinga2/etc/icinga2/pki/ganymed.csr'.
information/cli: Signing CSR with CA and writing certificate to '/tmp/icinga2/etc/icinga2/pki/ganymed.crt'.
information/pki: Writing certificate to file '/tmp/icinga2/etc/icinga2/pki/ganymed.crt'.
information/cli: Copying CA certificate to '/tmp/icinga2/etc/icinga2/pki/ca.crt'.
Generating master configuration for Icinga 2.
information/cli: Adding new ApiUser 'root' in '/tmp/icinga2/etc/icinga2/conf.d/api-users.conf'.
information/cli: Enabling the 'api' feature.
Enabling feature api. Make sure to restart Icinga 2 for these changes to take effect.
information/cli: Dumping config items to file '/tmp/icinga2/etc/icinga2/zones.conf'.
information/cli: Created backup file '/tmp/icinga2/etc/icinga2/zones.conf.orig'.
Please specify the API bind host/port (optional):
Bind Host []: 
Bind Port []: 
information/cli: Created backup file '/tmp/icinga2/etc/icinga2/features-available/api.conf.orig'.
information/cli: Updating constants.conf.
information/cli: Created backup file '/tmp/icinga2/etc/icinga2/constants.conf.orig'.
information/cli: Updating constants file '/tmp/icinga2/etc/icinga2/constants.conf'.
information/cli: Updating constants file '/tmp/icinga2/etc/icinga2/constants.conf'.
information/cli: Updating constants file '/tmp/icinga2/etc/icinga2/constants.conf'.
Should the wizard disable the inclusion of the conf.d directory [y/N]: y
information/cli: Updating configuration file '/tmp/icinga2/etc/icinga2/icinga2.conf'.
information/cli: Updating configuration file '/tmp/icinga2/etc/icinga2/icinga2.conf'.
Done.

Now restart your Icinga 2 daemon to finish the installation!
```

The updated icinga2.conf file looks this:
```
/**
 * Icinga 2 configuration file
 * - this is where you define settings for the Icinga application including
 * which hosts/services to check.
 *
 * For an overview of all available configuration options please refer
 * to the documentation that is distributed as part of Icinga 2.
 */

/**
 * The constants.conf defines global constants.
 */
include "constants.conf"

/**
 * The zones.conf defines zones for a cluster setup.
 * Not required for single instance setups.
 */
include "zones.conf"

/**
 * The Icinga Template Library (ITL) provides a number of useful templates
 * and command definitions.
 * Common monitoring plugin command definitions are included separately.
 */
include <itl>
include <plugins>
include <plugins-contrib>
include <manubulon>

/**
 * This includes the Icinga 2 Windows plugins. These command definitions
 * are required on a master node when a client is used as command endpoint.
 */
include <windows-plugins>

/**
 * This includes the NSClient++ check commands. These command definitions
 * are required on a master node when a client is used as command endpoint.
 */
include <nscp>

/**
 * The features-available directory contains a number of configuration
 * files for features which can be enabled and disabled using the
 * icinga2 feature enable / icinga2 feature disable CLI commands.
 * These commands work by creating and removing symbolic links in
 * the features-enabled directory.
 */
include "features-enabled/*.conf"

/**
 * The repository.d directory contains all configuration objects
 * managed by the 'icinga2 repository' CLI commands.
 */
include_recursive "repository.d"

/**
 * Although in theory you could define all your objects in this file
 * the preferred way is to create separate directories and files in the conf.d
 * directory. Each of these files must have the file extension ".conf".
 */
// include_recursive "conf.d"
include "conf.d/api-users.conf"
```

This should fix #4508

dnsmichi commented:
From a first look: It should be added to the `node setup` command too, i.e. --disable-confd as parameter. This is mandatory to add a check box to the Windows wizard which uses this cli command.

dnsmichi commented:
From a first code review, it looks good. Still need to test it in deep.

dnsmichi commented on lib/cli/nodesetupcommand.cpp:
```diff
@@ -443,6 +456,18 @@ int NodeSetupCommand::SetupNode(const boost::program_options::variables_map& vm,
+		if(apiUsersFile)
+			NodeUtility::UpdateConfiguration("\"conf.d/api-users.conf\"", true, false);
+		else
+			Log(LogWarning, "cli", "Included File dosn't exists " + apiUsersFilePath);
```
Typos here.

dnsmichi commented on lib/cli/nodesetupcommand.cpp:
```diff
@@ -226,6 +227,18 @@ int NodeSetupCommand::SetupMaster(const boost::program_options::variables_map& v
+		if(apiUsersFile)
+			NodeUtility::UpdateConfiguration("\"conf.d/api-users.conf\"", true, false);
+		else
+			Log(LogWarning, "cli", "Included File dosn't exists " + apiUsersFilePath);
```
Typos.

dnsmichi commented on lib/cli/nodeutility.cpp:
```diff
@@ -639,6 +639,69 @@ void NodeUtility::SerializeObject(std::ostream& fp, const Dictionary::Ptr& objec
+
+	String affectedInclude = value;
+
+	if(recursive)
```
Code style.

dnsmichi commented on lib/cli/nodeutility.cpp:
```diff
@@ -639,6 +639,69 @@ void NodeUtility::SerializeObject(std::ostream& fp, const Dictionary::Ptr& objec
+	}
+
+	if(include == true && found == false)
+	{
```
Code style.

dnsmichi replied:
Use `if (include && !found)` instead.

dnsmichi commented on lib/cli/nodeutility.cpp:
```diff
@@ -639,6 +639,69 @@ void NodeUtility::SerializeObject(std::ostream& fp, const Dictionary::Ptr& objec
+	ifp.close();
+	ofp.close();
+
+	if (rename(tempFile.CStr(), configurationFile.CStr()) < 0) {
```
Requires the _unlink ifdef for Windows beforehand.

dnsmichi commented on lib/cli/nodewizardcommand.cpp:
```diff
@@ -578,6 +579,28 @@ int NodeWizardCommand::Run(const boost::program_options::variables_map& vm,
+
+	choice = answer;
+
+	if(choice.Contains("y")) {
```
Style.

dnsmichi commented on lib/cli/nodewizardcommand.cpp:
```diff
@@ -578,6 +579,28 @@ int NodeWizardCommand::Run(const boost::program_options::variables_map& vm,
+		if(apiUsersFile)
+			NodeUtility::UpdateConfiguration("\"conf.d/api-users.conf\"", true, false);
+		else
+			Log(LogWarning, "cli", "Included File dosn't exists " + apiUsersFilePath);
```
Typos.

dnsmichi commented on lib/cli/nodeutility.cpp:
```diff
@@ -639,6 +639,69 @@ void NodeUtility::SerializeObject(std::ostream& fp, const Dictionary::Ptr& objec
+			else
+				ofp << line << "\n";
+		}
+	}
```
Bracket style in the while block above.

mcktr commented:
Just a reminder to update the docs:

https://github.com/Icinga/icinga2/blob/master/doc/06-distributed-monitoring.md
https://github.com/Icinga/icinga2/blob/master/doc/11-cli-commands.md

mcktr commented:
I implemented most of the required changes from the review. I still have two to do and I need to test the changes again.

When I am finished I will squash everything into one commit and update this PR.

dnsmichi commented:
Okidoki, thanks :)

dnsmichi commented:
Mh, I was thinking about this .. keep it on hold for now. If we remove the bottom mode in 2.8, new installs should always disable the inclusion of "conf.d" on client setups. We probably don't need a wizard option for that.

mcktr commented:
From a first thought this seems to make sense. Nevertheless I guess the function `NodeUtility::UpdateConfiguration` could be used to disable the inclusion of "conf.d" on a client.

I will fixing the one outstanding requested change in the function and then I will wait until a decisions is made. Hope this is okay so far. :)

dnsmichi commented:
Yep, the code parts make sense a lot. But still I wanted to not cause too much work time here, if we come up with a better solution. As you know, conf.d triggers checks on their own on clients which causes irritation. This is the main cause to fix ;)

Crunsher commented:
Any udate on this @dnsmichi ? I can take over reviewing and testing but need an answer on the question whether we wan't to disable conf.d per default. We also have to make sure the graphical windows wizard works with these changes (afaik it uses node setup in the background)

mcktr commented:
@Crunsher From the linked issue I thought, that a decision was made to disable conf.d by default.

I have made minor changes on this PR to ensure that the conf.d directory is always disable when running node wizard/node setup and on master setups the api-users.conf gets included by default.

I merged my previous commits into a single commit and my new changes to a separate commit, to distinct between previous changes and new changes. I also rebased the branch against the latest master.

Crunsher commented:
Fails to build on Windows. I think you got a bit too enthusiastic moving the lines around :D
```
1>C:\git\icinga2\lib\cli\nodeutility.cpp(318): error C2065: 'constantsFile': undeclared identifier
1>C:\git\icinga2\lib\cli\nodeutility.cpp(318): error C2228: left of '.CStr' must have class/struct/union
```

mcktr commented:
Was a minor mistake while I copied & pasted the _unlink statement from the other function. Should now build on Windows (In a quick test it did). 

The log messages and console output should be changed I guess, but don't know how much I should Log or output to the console for the user, only errors or all steps that the wizard are doing?

Crunsher commented:
The only non-log console output should be the questions we ask the user in the wizard. Log should output everything the node command changes as LogInformation, errors as well.

These are the bottom lines after running node wizard for a master on a vanilla Icinga 2 installation:
```
// include_recursive "conf.d"
include "conf.d/api-users.conf"
```

We do need api-users, but I don't think we should keep it in conf.d at then.
Other than that it looks good to me.

Crunsher commented:
After talking to @lippserd and @gunnarbeutner we decided to take this topic off the 2.8.0.
There where multiple proposals on how we handle dealing with the example config: The most promising one was moving it from `icinga2/conf.d` to `icinga/example.d`, making it easy to disable (maybe through the cli) while not touching possible existing user or api created files.

dnsmichi commented:
Sorry to let you hang on this. I thought I would have time for it last week, unfortunately it is the time where everyone gets sick. 

I agree on the fact that we should find a common solution for this, and not rush something here.

mcktr commented:
* rebased against the current master
* now uses the interactive CLI wizard to ask the user (default to disable)

    ```
    Do you want to disable the inclusion of the conf.d directory [Y/n]: 
    Disable the inclusion of the conf.d directory...
    Checking if api-users.conf exist...
    ```
    * Output messages may needs to be adjusted a bit. 
    * If the user re-run the node wizard and choose not to disable conf.d, but did it in the first run, how would we proceed with this? After the second run the conf.d directory is still disabled. 

* introduced `--dont-disable-confd` as new command option to `node setup`

    If not set the conf.d directory will be removed. We might want to have a different name for this command option. 

I would suggest that we extend the windows wizard in a separate follow-up PR when we finished this one.

dnsmichi commented:
If the user chooses to enable conf.d, and you detect it being disabled, I would just print manual instructions aka edit the icinga2.conf manually. I don't see the benefit of negating the logic here.

Basically what we want to for satellites and clients - disable conf.d by default. That's the main purpose, and users can "opt out" if they want.

mcktr commented:
Fixed the conflicts and squashed all commits into a single one. I tested it on Debian and Windows, it's working so far.

It's a bit tricky to detect, if the user re-runs the node wizard or not, so I can print instructions for enabling conf.d manually. Currently I don't know how to do this properly, maybe I need an impulse in the right direction here. If this is not possible without creating an extra function for it (guess this is a bit beyond the goal) I suggest that we document this and we leave it for it.

dnsmichi commented on lib/cli/nodewizardcommand.cpp:
```diff
@@ -788,12 +807,52 @@ int NodeWizardCommand::MasterSetup() const
+	
+	String constants_file = Application::GetSysconfDir() + "/icinga2/constants.conf";
+
+	NodeUtility::CreateBackupFile(constants_file);
```
Please use the new NodeUtility:: GetConstantsConfPath() function.

Crunsher commented:
The node wizard failed to disable con'd inclusion for my icinga2.conf
```
[...]
/**
 * The repository.d directory contains all configuration objects
 * managed by the 'icinga2 repository' CLI commands.
 */
include_recursive "repository.d"

//include "conf.d/single.conf"
//include "conf.d/api-users.conf"
/**
 * Although in theory you could define all your objects in this file
 * the preferred way is to create separate directories and files in the conf.d
 * directory. Each of these files must have the file extension ".conf".
 */
include_recursive "conf.d/"
```

We can't really guarantee it works and don't have to, but a small notice should be given.

Crunsher commented:
1. Disabling conf.d without a icinga2.conf creates an empty icinga2.conf
2. Fails with an edited icinga2.conf says, it was successful. The problem here was my line was:
```
include_recursive "conf.d/"
```
Note the slash at the end.

In most cases this will be used on stock installations, but we should be more careful if that's not the case. I.e. if the utility isn't sure it was successful, it should say so.

I'm going to take it from here, make the changes and add the option to the windows wizard. Full speed towards 2.9.0 :3

Crunsher commented:
Tested and works for me. On Windows there is no message about conf.d exclusion possibly failing, but I don't think this will be much of an issue.

mcktr commented:
When using `node setup` with `--disable-confd` we currently don't warn for a failed exclusion, is this on purpose? 

**Test**

```
# vim /tmp/icinga2/etc/icinga2/icinga2.conf

include_recursive "conf.d/"

# /tmp/icinga2/sbin/icinga2 node setup --master --disable-confd

information/cli: Checking in existing certificates for common name 'metis'...
information/cli: Certificates not yet generated. Running 'api setup' now.
information/cli: Generating new CA.
information/base: Writing private key to '/tmp/icinga2/var/lib/icinga2/ca//ca.key'.
information/base: Writing X509 certificate to '/tmp/icinga2/var/lib/icinga2/ca//ca.crt'.
information/cli: Generating new CSR in '/tmp/icinga2/var/lib/icinga2/certs//metis.csr'.
information/base: Writing private key to '/tmp/icinga2/var/lib/icinga2/certs//metis.key'.
information/base: Writing certificate signing request to '/tmp/icinga2/var/lib/icinga2/certs//metis.csr'.
information/cli: Signing CSR with CA and writing certificate to '/tmp/icinga2/var/lib/icinga2/certs//metis.crt'.
information/pki: Writing certificate to file '/tmp/icinga2/var/lib/icinga2/certs//metis.crt'.
information/cli: Copying CA certificate to '/tmp/icinga2/var/lib/icinga2/certs//ca.crt'.
information/cli: Generating master configuration for Icinga 2.
information/cli: Adding new ApiUser 'root' in '/tmp/icinga2/etc/icinga2/conf.d/api-users.conf'.
information/cli: Enabling the 'api' feature.
Enabling feature api. Make sure to restart Icinga 2 for these changes to take effect.
information/cli: Generating zone and object configuration.
information/cli: Dumping config items to file '/tmp/icinga2/etc/icinga2/zones.conf'.
information/cli: Created backup file '/tmp/icinga2/etc/icinga2/zones.conf.orig'.
information/cli: Updating the APIListener feature.
information/cli: Created backup file '/tmp/icinga2/etc/icinga2/features-available/api.conf.orig'.
information/cli: Updating 'NodeName' constant in '/tmp/icinga2/etc/icinga2/constants.conf'.
information/cli: Created backup file '/tmp/icinga2/etc/icinga2/constants.conf.orig'.
information/cli: Updating 'ZoneName' constant in '/tmp/icinga2/etc/icinga2/constants.conf'.
information/cli: Backup file '/tmp/icinga2/etc/icinga2/constants.conf.orig' already exists. Skipping backup.
information/cli: Updating 'TicketSalt' constant in '/tmp/icinga2/etc/icinga2/constants.conf'.
information/cli: Backup file '/tmp/icinga2/etc/icinga2/constants.conf.orig' already exists. Skipping backup.
information/cli: Edit the api feature config file '/tmp/icinga2/etc/icinga2/features-available/api.conf' and set a secure 'ticket_salt' attribute.
information/cli: Updating' "conf.d"' include in '/tmp/icinga2/etc/icinga2/icinga2.conf'.
information/cli: Created backup file '/tmp/icinga2/etc/icinga2/icinga2.conf.orig'.
information/cli: Updating' "conf.d/api-users.conf"' include in '/tmp/icinga2/etc/icinga2/icinga2.conf'.
information/cli: Backup file '/tmp/icinga2/etc/icinga2/icinga2.conf.orig' already exists. Skipping backup.
information/cli: Make sure to restart Icinga 2.

# cat /tmp/icinga2/etc/icinga2/icinga2.conf                       

[...]

include_recursive "conf.d/"
include "conf.d/api-users.conf"

```

dnsmichi commented:
https://github.com/Icinga/icinga2/pull/5426/files#diff-a68014714a3ecf2821c9a319821b4bfaR817 should be changed into NodeUtility::GetConstantsConfPath().

Crunsher commented:
> When using node setup with --disable-confd we currently don't warn for a failed exclusion, is this on purpose?

Done

> changed into NodeUtility::GetConstantsConfPath()

And done

dnsmichi commented:
Some review remarks, commits of mine fix them.

- Code style with conditions, e.g. `if(...`
- Indent for api-users.conf inclusion
- Ternary operators for assignments are bad style, just use a simple if-then-else which is readable
- `const bool& recursive` doesn't make sense
- Wording and grammar of log messages
- Use `Utility::PathExists()` instead of an ifstream (which isn't closed anymore, may cause troubles)
- path to api-users.conf is used in multiple locations, should be moved into a getter in a utility helper
- Readability: if-then-else with more than 1 line body should always have brackets

- node wizard does it correctly, to not include api-users on satellite/clients (deployed in a different fashion). node setup is incorrect, it adds it again. 

- Commenting out and adding something should have an auto-generation comment

```
// Disabled by the node setup CLI command on 2018-05-08 15:44:31 +0200
// include_recursive "conf.d"

// Added by the node setup CLI command on 2018-05-08 15:44:31 +0200
include "conf.d/api-users.conf"
```

- Documentation for the CLI commands and node wizard CLI output is entirely missing. The screenshot update ticket is separate from this one.

I've rebased it against master, added my commits and won't squash it as there's three people involved here. Only amended the commit messages to reference the other issue too.
